### PR TITLE
Recipe App

### PR DESCRIPTION
**Pull request by cvalencia1991**
# In this milestone @eng-mohamed-eldeeb  and @cvalencia1991 and @eng-mohamed-eldeeb  did the following activities:

- Use Ruby gems as a software package system.
- Install Ruby on Rails framework.
- Understand Rails RESTful design and router.
- Use controllers to handle requests and render empty views.
- Understand Rails naming conventions.
- Use params from browser requests in a safe way.
- Use preprocessed HTML files with embedded Ruby code.
- Use layouts and templates for shared content.
- Use database migration files to maintain database schema.
- Use validations for models.
- Secure app from n+1 problems.
- Understand what ORM is.
- Set up associations between models.
- Build a web app that requires the user to log in.
- Describe the difference between authorization and authentication.
- Use Devise gem for authentication.
- Understand how sessions and cookies can support authentication.
- Limit access to web app resources based on authorization rules.

### CODE REVIEW

File: app/views/foods/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+         <% if user_signed_in? %>
+                  <th>
+                    <%= link_to "Remove",user_recipe_food_path(params[:user_id], params[:recipe_id], food.id),data: {turbo_method: :delete}, class:"p-4 font-medium hover:text-red-500"  %>
+                    <%= link_to "Modify",edit_user_recipe_path(params[:user_id], params[:recipe_id], food.id), method: :edit, class:"p-4 font-medium hover:text-green-400"  %>
+                  </th>
+</tr>
+            <% end %>
+      <% end %>
+
+      </table>
```
Comment (danmainah) [inline app/views/foods/index.html.erb:27]: - [ ] kindly ensure your design matches the design given, the food table should be below the add food button check the comparison below

Wireframe Design          |  Your design     
:-------------------------:|:-------------------------:
image  |  image

File: app/views/recipe_foods/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+<div id="recipe_foods">
+  <% @recipe_foods.each do |recipe_food| %>
+    <%= render recipe_food %>
+    <p>
+    <p class="my-7 border-solid w-56 rounded-md border-2 borde  r-slate-200 p-4 bg-gray-50 hover:bg-black hover:text-white text-center">
+      <%= link_to "Show this recipe food", user_recipe_recipe_foods_path %>
+    </p>
+  <% end %>
+</div>
```
Comment (danmainah) [inline app/views/recipe_foods/index.html.erb:13]: kindly ensure your design matches the design given, check the comparison below to see the differences 

Wireframe Design          |  Your design     
:-------------------------:|:-------------------------:
image |  image

File: app/views/recipe_foods/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<p style="color: green"><%= notice %></p>
+
+<%= render @recipe_food %>
+
+<div>
+  <%= link_to "Edit this recipe food", edit_recipe_food_path(@recipe_food) %> |
+  <%= link_to "Back to recipe foods", recipe_foods_path %>
+
+  <%= button_to "Destroy this recipe food", @recipe_food, method: :delete %>
```
Comment (danmainah) [inline app/views/recipe_foods/show.html.erb:9]: - [ ] kindly ensure your design matches the design given, check the comparison below to see the differences 

Wireframe Design          |  Your design     
:-------------------------:|:-------------------------:
image |  image

File: app/views/public_recipe/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+            <div class="bg-white border-b-2 text-slate-700 border-slate-700">
+                <h3 class="p-4 text-lg">user name: <%= user.name%></h3>
+                <h4 class="p-4 text-xl font-medium">the recipe is: <%= recipe.name %></h1>
+                <p class="my-7 border-solid w-56 rounded-md border-2 borde  r-slate-200 p-4 bg-gray-50 hover:bg-black hover:text-white text-center">
+                    <%= link_to "recipe details", user_recipe_path(user.id,recipe.id) %>
+                </p>
+            </div>
+        <%end%>
+    <%end%>
+<%end%>
```
Comment (danmainah) [inline app/views/public_recipe/index.html.erb:20]: - [ ] kindly ensure your design matches the wireframe design given , below is a comparison to see the differences

Wireframe Design          |  Your design     
:-------------------------:|:-------------------------:
image  |  image

File: config/routes.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+Rails.application.routes.draw do
+  root 'public_recipe#index'
+  devise_for :users, path_names: { sign_in: 'login', sign_out: 'logout', password: 'secret', confirmation: 'verification', unlock: 'unblock', registration: 'register', sign_up: 'cmon_let_me_in' }
+  resources :users do
+    resources :recipes do
+      resources :recipe_foods
+      resources :foods
+      resources :shopinglists
+    end
+  end
```
Comment (danmainah) [inline config/routes.rb:10]: - [ ] @cvalencia1991 and @eng-mohamed-eldeeb kindly ensure you only display routes that you are using currently I can see routes that are not in use.

File: app/controllers/recipe_foods_controller.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  end
+
+  # DELETE /recipe_foods/1 or /recipe_foods/1.json
+  def destroy
+    @recipe_food.destroy
+
+    respond_to do |format|
+      format.html { redirect_to recipe_foods_url, notice: 'Recipe food was successfully destroyed.' }
+      format.json { head :no_content }
+    end
```
Comment (danmainah) [inline app/controllers/recipe_foods_controller.rb:56]: - [ ] @team kindly create a fail path/outcome for your destroy method like you did in your create and update methods

File: app/controllers/foods_controller.rb [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    end
+  end
+
+  # DELETE /foods/1 or /foods/1.json
+  def destroy
+    @food.destroy
+    respond_to do |format|
+      format.html { redirect_to user_recipe_path, notice: 'Food was successfully destroyed.' }
+      format.json { head :no_content }
+    end
```
Comment (danmainah) [inline app/controllers/foods_controller.rb:55]: - [ ] @team kindly create a fail path/outcome for your destroy method like you did in your create and update methods

File: app/controllers/recipes_controller.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  # DELETE /recipes/1 or /recipes/1.json
+  def destroy
+    @recipe.destroy
+
+    respond_to do |format|
+      format.html { redirect_to user_recipes_path(params[:user_id]), notice: 'Recipe was successfully destroyed.' }
+      format.json { head :no_content }
+    end
```
Comment (danmainah) [inline app/controllers/recipes_controller.rb:59]: @team kindly create a fail path/outcome for your destroy method like you did in your create and update methods

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @cvalencia1991 and @eng-mohamed-eldeeb , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

✔️ No linter errors
✔️ Gitflow was correctly used
✔️  Good design


# Required Changes ♻️

- [x] @team there is a bug in your application, when a user logs in initially before adding any recipes and clicks on all recipes I get the notification below, kindly take a look at it.
image


_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/controllers/foods_controller.rb (hunk 7 above)
Comment (cvalencia1991) [reply app/controllers/foods_controller.rb:55]: Ok you right i fix it

File: app/views/recipes/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,37 @@
+<div class="w-[80%] flex flex-col items-center">
+
+<%= render @recipe %>
+<div class="flex gap-5">
+    <%= link_to "add ingredeints", new_user_recipe_recipe_food_path(params[:user_id],params[:id]), class:"border-slate-700 rounded-md border-2 p-2 bg-gray-50 text-slate-700 hover:bg-black hover:text-white text-center mb-4"%>
+    <%= link_to "general shooping list", user_recipe_shopinglists_path(params[:user_id],params[:id]), class:"border-slate-700 rounded-md border-2 p-2 bg-gray-50 text-slate-700 hover:bg-black hover:text-white text-center mb-4"%>
+</div>
+    <% @recipefoods.each do |foods| %>
+    sup
```
Comment (franclobo) [inline app/views/recipes/show.html.erb:9]: - [ ] While you were correcting this section you accidentally added 👀 **sup** word, please remove it for a better user interface.

File: app/views/public_recipe/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+                    <h2 class="my-5 border-solid w-72 rounded-md p-2 bg-black text-white text-xl text-center">amought of items to buy: <%= @foods.length%></h2>
+                    <h3 class="my-5 border-solid w-80 rounded-md p-2 bg-black text-white text-xl text-center">total value of food needed it:
+                    <%=  food.quantity %></h3>
+                </div>
+                    <p class="my-7 border-solid w-56 rounded-md border-2 borde  r-slate-200 p-4 bg-gray-50 hover:bg-black hover:text-white text-center">
+                        <%= link_to "recipe details", user_recipe_path(food.user.id,recipe.id) %>
+                    </p>
+        </div>
+        <%end%>
+    <%end%>
```
Comment (franclobo) [inline app/views/public_recipe/index.html.erb:23]: - [ ] As required by the previous code reviewer please 🙏🏽 match with the wireframe, kindly align the content at the center of the page.

**Review by franclobo — CHANGES_REQUESTED** [review]
Hi team, 

# ♻️ Required changes ♻️ 

so-close-jim-halpert

Great work on making the changes 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## Highlights

- Good use of Bootstrap ✔️ 
- Correct authentication with devise ✔️ 
- Gem cancancan is used of authorization ✔️ 
## Suggested changes
_Check the comments under the review._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: app/views/recipe_foods/index.html.erb [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+<h1 class="mb-10 text-3xl font-medium">Recipe foods</h1>
+
+<div id="recipe_foods">
+  <% @recipe_foods.each do |recipe_food| %>
+    <%= render recipe_food %>
+    <p>
+    <p class="my-7 border-solid w-56 rounded-md border-2 borde  r-slate-200 p-4 bg-gray-50 hover:bg-black hover:text-white text-center">
+      <%= link_to "Show this recipe food", user_recipe_recipe_foods_path %>
+    </p>
+
```
[the remark this answers is not in the document]
Comment (cvalencia1991) [reply app/views/recipe_foods/index.html.erb:12]: ok, I understand that the problem is in  **My recipes**, so in the next review to see the  issue  you can check the next image that the problem is in the styles of the page 

image

**Review by codepantha — APPROVED** [review]
# STATUS: APPROVED 🟢 ✔️ 

## Hi team 👋, 

Well done for implementing the required changes 👏 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by cvalencia1991** [comment]
thank you 😄

**Pull request merged**